assets/js/components/ChecksResults/checksUtils.js

```diff
@@ -32,31 +34,71 @@ export const getCheckResults = (executionData) => {
 };
 
 export const getHosts = (checkResults) =>
-  checkResults.flatMap(({ agents_check_results }) =>
-    agents_check_results.map(({ agent_id }) => agent_id)
+  uniq(
+    checkResults.flatMap(({ agents_check_results }) =>
+      agents_check_results.map(({ agent_id }) => agent_id)
+    )
   );
 
 export const getChecks = (checkResults) =>
   checkResults.map(({ check_id }) => check_id);
 
-export const getHealth = (checkResults, checkID, agentID) => {
+export const getCheckHealthByAgent = (checkResults, checkID, agentID) => {
+  if (!checkResults) {
+    return {};
+  }
+
   const checkResult = checkResults.find(({ check_id }) => check_id === checkID);
   if (!checkResult) {
-    return undefined;
+    return {};
   }
 
   const agentCheckResult = checkResult.agents_check_results.find(
     ({ agent_id }) => agent_id === agentID
   );
 
-  const failedExpectationEvaluations = agentCheckResult?.expectation_evaluations
-    .filter((expectationEvaluation) => 'message' in expectationEvaluation)
-    .filter(({ type }) => type !== 'expect');
+  if (!agentCheckResult) {
+    return {};
+  }
+
+  // agentCheckError
+  if (agentCheckResult?.type) {
+    return {
+      health: 'critical',
+      error: agentCheckResult.message,
+    };
+  }
+
+  // expectation evaluation error , malformed expression most probably
+  const evaluationErrors = agentCheckResult?.expectation_evaluations.filter(
+    ({ message }) => message
+  ).length;
+
+  // expect evaluating to false
+  const failedExpectEvaluations =
+    agentCheckResult?.expectation_evaluations.filter(
+      ({ type, return_value: returnValue }) => type === 'expect' && !returnValue
+    ).length;
+
+  // expect_same
+  const failedExpectSameEvaluations =
+    agentCheckResult?.expectation_evaluations.filter(
+      ({ name, type }) =>
+        type === 'expect_same' &&
+        !checkResult.expectation_results.find(
+          ({ name: resultName }) => resultName === name
+        )?.result
+    ).length;
+
+  const failedExpectations =
+    evaluationErrors + failedExpectEvaluations + failedExpectSameEvaluations;
+
+  const health = failedExpectations > 0 ? checkResult.result : 'passing';
```

Comment:
What about relying on wanda's response for the health :thinking: 

And since this has been introduced https://github.com/trento-project/wanda/pull/104 (counts for passing/warning/critical checks), could also those be used in the frontend?

Reply:
We cannot. The way the Frontend is now, we need to know the result for each host, which wanda doesn't store, as it only stores the complete check result.
Once we change this view, using a more checks result oriented rather than host result, we will need to revisit this whole thing

Reply:
Alright, got it. Thanks!

Reply:
Yeah, @arbulu89 is right, the counters are for checks results, we need to count single expectations too :smile: